src/server_manager/ui_components/outline-validated-input.html

```diff
@@ -0,0 +1,190 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+<!-- outline-validated-input
+    This is an input, with a cancel and a save button, which performs client-side validation and has an
+    event-based hook for server-side evaluation.
+
+    Attributes:
+        * name: A unique name used to build element id's.  Default: null
+        * editable: Is this input editable?  Default: false
+        * visible: Is this input visible?  Default: false
+        * label: The label for the input.  Default: null
+        * value: The initial value entered in the input.  SHOULD be a valid value.  Default: null
+        * allowed-pattern: Regex describing what inputs are allowed.  Users will be prevented from
+            entering inputs which don't follow the pattern.  Default: ".*"
+        * max-length: The number of characters allowed in the input.  Default: Number.POSITIVE_INFINITY
+        * client-side-validator: A function which takes a string and returns either an empty string on success
+            or an error message on failure.  This function will be called on every keystroke. Default: () => ""
+        * event: The name of the event fired when the save button is tapped.  Passes the
+            current input value as "value" and the Polymer object as "ui".  The handler for this event
+            MUST call ui.setSavedState() or ui.setErrorState(message) depending on its result. Default: null
+ -->
+
+<link rel='import' href='./cloud-install-styles.html'>
+<link rel='import' href='./outline-server-settings-styles.html'>
+
+<dom-module id='outline-validated-input'>
+    <template>
+        <style include='cloud-install-styles'></style>
+        <style include='outline-server-settings-styles'></style>
+        <style>
+            [id$=Container] {
+                display: flex;
+                justify-content: space-between;
+                align-items: center;
+                max-width: 545px;
+            }
+
+            [id$=SaveButton] {
+                background: var(--primary-green);
+                color: #fff;
+                height: 36px;
+                width: 83px;
+            }
+
+            [id$=SaveButton][disabled] {
+                background-color: rgba(255, 255, 255, .12);
+                color: rgba(255, 255, 255, .3);
+            }
+
+            [id$=CancelButton] {
+                color: #fff;
+                height: 36px;
+                width: 83px;
+            }
+
+            [id$=CancelButton][disabled] {
+                color: rgba(255, 255, 255, 0.3);
+                background-color: rgba(0, 0, 0, 0);
+            }
+        </style>
+        <div id="[[name]]Container" hidden$="[[!visible]]">
```

Comment:
Unless I'm missing something, it shouldn't be necessary to prepend the name attribute to these IDs.

Reply:
If we make more than one outline-validated-input, then we would have multiple elements with the same id

Reply:
Spoke offline -- clarified that shadow doms introduce their own scope